Provide API to preserve source range and callstack information during graph rewrite

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58492 [Pytorch] Update FuseLinear to map source range information
* **#58300 [Pytorch] Provide API to preserve source range and callstack information during graph rewrite**

Summary:
Current state: During graph rewriting that can fuse nodes or add nodes
result in new nodes without debug information that was available in
original node. Thus we lose this information during graph rewrite.

This PR changes graph rewriting API to let user specify how the values
in the replacement pattern map to values in the pattern to be matched.
Then the graph rewriting will copy source range and inlined callstack
from the matched nodes onto the nodes being inserted.

Test Plan:
python test/test_jit.py
TestJit.test_pattern_based_rewrite_with_source_range_preserved

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28512465](https://our.internmc.facebook.com/intern/diff/D28512465)

kimishpatel committed:
[Pytorch] Provide API to preserve source range and callstack information during graph rewrite

Summary:
Current state: During graph rewriting that can fuse nodes or add nodes
result in new nodes without debug information that was available in
original node. Thus we lose this information during graph rewrite.

This PR changes graph rewriting API to let user specify how the values
in the replacement pattern map to values in the pattern to be matched.
Then the graph rewriting will copy source range and inlined callstack
from the matched nodes onto the nodes being inserted.

Test Plan:
python test/test_jit.py
TestJit.test_pattern_based_rewrite_with_source_range_preserved

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/subgraph_rewrite.cpp:
```diff
@@ -17,13 +43,16 @@ graph(%x, %w, %b):
 graph(%x, %w, %b):
   %r = aten::convrelu(%x, %w, %b)
-  return (%r))IR");
+  return (%r))IR",
```
Do we use the default patterns anywhere btw? I added them in the beginning, but I suspect we register patterns explicitly in all use cases - if that's the case we could probably nuke this.

ZolotukhinM commented on torch/csrc/jit/passes/subgraph_rewrite.cpp:
```diff
@@ -56,7 +85,26 @@ void SubgraphRewriter::rewriteSinglePatternOnGraph(
+
+  // First construct map of Node*-to-Node*
+  // This maps Nodes in replacement graph to nodes in pattern graph
+  // given the value_name_map, which maps value names from repalcement
```
Typo: repalcment

ZolotukhinM reviewed (approved):
LGTM!

kimishpatel committed:
Update on "[Pytorch] Provide API to preserve source range and callstack information during graph rewrite"


Summary:
Current state: During graph rewriting that can fuse nodes or add nodes
result in new nodes without debug information that was available in
original node. Thus we lose this information during graph rewrite.

This PR changes graph rewriting API to let user specify how the values
in the replacement pattern map to values in the pattern to be matched.
Then the graph rewriting will copy source range and inlined callstack
from the matched nodes onto the nodes being inserted.

Test Plan:
python test/test_jit.py
TestJit.test_pattern_based_rewrite_with_source_range_preserved

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28512465).

kimishpatel committed:
Update on "[Pytorch] Provide API to preserve source range and callstack information during graph rewrite"


Summary:
Current state: During graph rewriting that can fuse nodes or add nodes
result in new nodes without debug information that was available in
original node. Thus we lose this information during graph rewrite.

This PR changes graph rewriting API to let user specify how the values
in the replacement pattern map to values in the pattern to be matched.
Then the graph rewriting will copy source range and inlined callstack
from the matched nodes onto the nodes being inserted.

Test Plan:
python test/test_jit.py
TestJit.test_pattern_based_rewrite_with_source_range_preserved

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28512465](https://our.internmc.facebook.com/intern/diff/D28512465)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28512465).

kimishpatel committed:
Update on "[Pytorch] Provide API to preserve source range and callstack information during graph rewrite"


Summary:
Current state: During graph rewriting that can fuse nodes or add nodes
result in new nodes without debug information that was available in
original node. Thus we lose this information during graph rewrite.

This PR changes graph rewriting API to let user specify how the values
in the replacement pattern map to values in the pattern to be matched.
Then the graph rewriting will copy source range and inlined callstack
from the matched nodes onto the nodes being inserted.

Test Plan:
python test/test_jit.py
TestJit.test_pattern_based_rewrite_with_source_range_preserved

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28512465](https://our.internmc.facebook.com/intern/diff/D28512465)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28512465).
